In [2]:
import os
import pandas as pd
pd.set_option('mode.chained_assignment',  None)
pd.options.display.max_rows = 200

In [3]:
os.chdir('C://Users//nhw85//OneDrive//바탕 화면//정보')


new_df = pd.DataFrame()
for file_name in os.listdir('C://Users//nhw85//OneDrive//바탕 화면//정보'):
    abspath = os.path.abspath(file_name)
    file = pd.read_csv(abspath, index_col = 0)
    new_df = pd.concat([new_df, file], axis = 0, ignore_index=False)
    

In [4]:
# new_df['team'] = new_df.index
renew_df = new_df.reset_index()
renew_df
renew_df = renew_df.drop(['player_number'], axis = 1)
renew_df = renew_df.dropna(subset=['player_name'], axis = 0)
renew_df

player_team player_name pitcher_batting  birth_date height_weight
1             HH         최태원            우투우타  1970-08-19   179cm, 83kg
2             HH         이철성            우투우타  1957-03-14   170cm, 72kg
3             HH         김광수            우투우타  1959-01-03   165cm, 80kg
4             HH         이상군            우투우타  1962-04-21   176cm, 72kg
5             HH        나카시마            우투우타  1962-07-27   188cm, 92kg
...          ...         ...             ...         ...           ...
2566          WO         이병규            우투좌타  1994-10-05   175cm, 73kg
2567          WO         김준완            우투좌타  1991-01-20   174cm, 73kg
2568          WO         이용규            좌투좌타  1985-08-26   170cm, 74kg
2569          WO         이형종            우투우타  1989-06-07   183cm, 87kg
2570          WO         이정후            우투좌타  1998-08-20   185cm, 86kg

[2563 rows x 5 columns]

In [5]:
renew_df['pitcher_batting'].isnull().sum()

4

In [6]:
batting_hand = []
throwing_hand = []

for ifo in renew_df['pitcher_batting']:
    try:
        throwing_hand.append(ifo[:2])
        batting_hand.append(ifo[2:])
    except:
        throwing_hand.append('작성필요')
        batting_hand.append('작성필요')
        

renew_df['batting_hand'] = batting_hand
renew_df['throwing_hand'] = throwing_hand

# pitcher_batting 컬럼삭제
renew_df = renew_df.drop(['pitcher_batting'], axis = 1)

####
renew_df



player_team player_name  birth_date height_weight batting_hand  \
1             HH         최태원  1970-08-19   179cm, 83kg           우타   
2             HH         이철성  1957-03-14   170cm, 72kg           우타   
3             HH         김광수  1959-01-03   165cm, 80kg           우타   
4             HH         이상군  1962-04-21   176cm, 72kg           우타   
5             HH        나카시마  1962-07-27   188cm, 92kg           우타   
...          ...         ...         ...           ...          ...   
2566          WO         이병규  1994-10-05   175cm, 73kg           좌타   
2567          WO         김준완  1991-01-20   174cm, 73kg           좌타   
2568          WO         이용규  1985-08-26   170cm, 74kg           좌타   
2569          WO         이형종  1989-06-07   183cm, 87kg           우타   
2570          WO         이정후  1998-08-20   185cm, 86kg           좌타   

     throwing_hand  
1               우투  
2               우투  
3               우투  
4               우투  
5               우투  
...            ...  
2566            우투  
2567            우투  
2568            좌투  
2569            우투  
2570            우투  

[2563 rows x 6 columns]

In [7]:
renew_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2563 entries, 1 to 2570
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   player_team    2563 non-null   object
 1   player_name    2563 non-null   object
 2   birth_date     2563 non-null   object
 3   height_weight  2563 non-null   object
 4   batting_hand   2563 non-null   object
 5   throwing_hand  2563 non-null   object
dtypes: object(6)
memory usage: 140.2+ KB


In [8]:
renew_df.describe()

player_team player_name  birth_date height_weight batting_hand  \
count         2563        2563        2563          2563         2563   
unique          10         935         930           486            4   
top             KT         김상수  1971-10-30   185cm, 90kg           우타   
freq           259          13          11            39         1685   

       throwing_hand  
count           2563  
unique             5  
top               우투  
freq            2027

In [9]:
player_name_list = []

for name in renew_df['player_name']:
    player_name_list.append(name)
    
print(player_name_list) # 1294


['최태원', '이철성', '김광수', '이상군', '나카시마', '윤재국', '신경현', '김정준', '심수창', '이동걸', '박정진', '송신영', '송창식', '장민재', '비야누에바', '안영명', '송은범', '오간도', '윤규진', '정우람', '차일목', '조인성', '임익준', '송광민', '정근우', '신성현', '강경학', '로사리오', '김주현', '김태균', '이시원', '이양기', '최진행', '장민석', '김원석', '한용덕', '송진우', '장종훈', '전형도', '김해님', '채종국', '고동진', '강인권', '이양기', '심수창', '강승현', '김범수', '송창식', '박주홍', '휠러', '송은범', '샘슨', '서균', '윤규진', '정우람', '박상원', '최재훈', '지시완', '최윤석', '정경운', '오선진', '송광민', '정근우', '하주석', '김태균', '이시원', '이용규', '양성우', '호잉', '백진우', '강상원', '한용덕', '송진우', '장종훈', '전형도', '채종국', '타나베', '고동진', '차일목', '이양기', '김범수', '김재영', '이태양', '장민재', '안영명', '서폴드', '송은범', '채드벨', '서균', '정우람', '박상원', '임준섭', '최재훈', '지시완', '하주석', '오선진', '송광민', '정근우', '김태연', '노시환', '정은원', '이성열', '김태균', '양성우', '호잉', '김민하', '장진혁', '장종훈', '김성래', '전형도', '채종국', '정민태', '정현석', '고동진', '차일목', '신정락', '임준섭', '김범수', '이태양', '장시환', '장민재', '김진영', '안영명', '이현호', '서폴드', '김민우', '정우람', '박상원', '최재훈', '이해창', '오선진', '송광민', '노시환', '하주석', '김회성', '정은원', '이성열', '김태균', '장진혁', '호잉', '장운호', '이용규', '정진호', '

In [10]:
os.chdir('C://Users//nhw85//KBO_crawler//data//final')

total_KBO_final = pd.DataFrame()
for file in os.listdir():
    if not (".csv" in file):
        continue
    abspath = os.path.abspath(file)
    temp_df = pd.read_csv(abspath, index_col = 0)
    
    total_KBO_final = pd.concat([total_KBO_final, temp_df], axis = 0, ignore_index=True) 

In [11]:
total_KBO_final

gameID pitcher batter  pitcher_ID  batter_ID  \
0       20170331HHOB02017     니퍼트    김원석       61240      62768   
1       20170331HHOB02017     니퍼트    장민석       61240      71347   
2       20170331HHOB02017     니퍼트    김태균       61240      71752   
3       20170331HHOB02017     니퍼트   로사리오       61240      66740   
4       20170331HHOB02017   비야누에바    민병헌       67742      76249   
...                   ...     ...    ...         ...        ...   
226638  20201031NCHT02020     박진우    이진영       63991      66657   
226639  20201031NCHT02020     박진우    유민상       63991      62265   
226640  20201031NCHT02020     박진우    이우성       63991      63260   
226641  20201031NCHT02020     박진우    한승택       63991      63722   
226642  20201031NCHT02020     박진우    최정용       63991      65464   

       pitch_result_x pa_result pa_result_detail        pitch_result_y  \
0                  타격        안타               안타                    타격   
1                  타격      희생번트             희생번트                    타격   
2                   볼        볼넷               볼넷            볼 볼 파울 볼 볼   
3                  타격     필드 아웃        라인드라이브 아웃            헛스윙 헛스윙 타격   
4                 헛스윙        삼진               삼진  볼 볼 스트라이크 헛스윙 파울 헛스윙   
...               ...       ...              ...                   ...   
226638            헛스윙        삼진               삼진   스트라이크 스트라이크 볼 볼 헛스윙   
226639              볼        볼넷               볼넷               볼 볼 볼 볼   
226640              볼        볼넷               볼넷         볼 볼 볼 스트라이크 볼   
226641             타격     포스 아웃            땅볼 아웃            헛스윙 볼 볼 타격   
226642             타격      내야안타             내야안타              볼 헛스윙 타격   

        sequence_id  
0                 1  
1                 2  
2                 3  
3                 4  
4                 5  
...             ...  
226638           79  
226639           80  
226640           81  
226641           82  
226642           83  

[226643 rows x 10 columns]

In [ ]:
total_KBO_final['batter']


In [17]:
# pitcher의 player_ID 컬럼 추가
df_1 = pd.merge(total_KBO_final, renew_df, left_on='pitcher', right_on='player_name', how='left')
df_1= df_1[['pitcher','pitcher_ID','birth_date', 'height_weight', 'batting_hand','throwing_hand']].drop_duplicates()
df_1.isnull().sum()

# batter의 player_ID 컬럼 추가
df_2= pd.merge(total_KBO_final, renew_df, left_on='batter', right_on='player_name', how='left')
df_2= df_2[['batter','batter_ID','birth_date', 'height_weight', 'batting_hand','throwing_hand']].drop_duplicates()
df_2.isnull().sum()

batter             0
batter_ID          0
birth_date       121
height_weight    121
batting_hand     121
throwing_hand    121
dtype: int64

In [18]:
df_1 = df_1.reset_index().drop(['index'], axis=1)
df_1.rename(columns = {'pitcher_ID':'player_id', 'pitcher':'player'}, inplace=True)
df_1

player  player_id  birth_date height_weight batting_hand throwing_hand
0      니퍼트      61240  1981-05-06  203cm, 103kg           우타            우투
1    비야누에바      67742  1983-11-28  188cm, 100kg           우타            우투
2      송창식      74731  1985-03-25  183cm, 104kg           우타            우투
3      박정진      99737  1976-05-27   183cm, 88kg           좌타            좌투
4      장민재      79764  1990-03-19   184cm, 98kg           우타            우투
..     ...        ...         ...           ...          ...           ...
557    류진욱      65949  1996-10-10   189cm, 88kg           우타            우투
558    홍원표      50430         NaN           NaN          NaN           NaN
559    김태현      67956         NaN           NaN          NaN           NaN
560    김태경      50992         NaN           NaN          NaN           NaN
561    안인산      50901         NaN           NaN          NaN           NaN

[562 rows x 6 columns]

In [19]:
df_2 = df_2.reset_index().drop(['index'], axis=1)
df_2.rename(columns = {'batter_ID':'player_id', 'batter':'player'}, inplace=True)
df_2

player  player_id  birth_date height_weight batting_hand throwing_hand
0      김원석      62768  1989-10-29   182cm, 83kg           우타            우투
1      장민석      71347  1982-05-09   180cm, 73kg           좌타            좌투
2      김태균      71752  1982-05-29  185cm, 110kg           우타            우투
3      김태균      71752  1971-08-19   176cm, 82kg           우타            우투
4     로사리오      66740  1989-02-23  180cm, 100kg           우타            우투
..     ...        ...         ...           ...          ...           ...
604    문상인      67063         NaN           NaN          NaN           NaN
605    류효승      50868         NaN           NaN          NaN           NaN
606    박시원      50996         NaN           NaN          NaN           NaN
607    허관회      69706         NaN           NaN          NaN           NaN
608    이호연      68504  1995-06-03   177cm, 87kg           좌타            우투

[609 rows x 6 columns]

In [20]:
personal_player_info = pd.concat([df_1, df_2],axis=0, ignore_index=True) # 총 1171명의 선수정보 2017-2023
personal_player_info

player  player_id  birth_date height_weight batting_hand throwing_hand
0       니퍼트      61240  1981-05-06  203cm, 103kg           우타            우투
1     비야누에바      67742  1983-11-28  188cm, 100kg           우타            우투
2       송창식      74731  1985-03-25  183cm, 104kg           우타            우투
3       박정진      99737  1976-05-27   183cm, 88kg           좌타            좌투
4       장민재      79764  1990-03-19   184cm, 98kg           우타            우투
...     ...        ...         ...           ...          ...           ...
1166    문상인      67063         NaN           NaN          NaN           NaN
1167    류효승      50868         NaN           NaN          NaN           NaN
1168    박시원      50996         NaN           NaN          NaN           NaN
1169    허관회      69706         NaN           NaN          NaN           NaN
1170    이호연      68504  1995-06-03   177cm, 87kg           좌타            우투

[1171 rows x 6 columns]

In [21]:
personal_player_info.isnull().sum() # null 값 유무확인

player             0
player_id          0
birth_date       287
height_weight    287
batting_hand     287
throwing_hand    287
dtype: int64

In [22]:
os.getcwd()

'C:\\Users\\nhw85\\KBO_crawler\\data\\final'

In [23]:
# personal_player_info.to_csv('personal_player_info' + '2017_2023' + '.csv', encoding = 'utf-8') 

- 선수 각 개인의 정보를 전처리 완료
- name 이나 player_id가 null 값인 것은 최종적으로 없었음
- 하지만 총 1171명의 선수중 birth_date, height_weight, batting_hand, throwing_hand 의 null 값이 동일하게 287개가 발견되었다.    

#### Salary.csv

In [24]:
os.chdir('C://Users//nhw85//Downloads//[KBO] 팀별 선수 연봉(17-23)')

player_salary = pd.DataFrame()
for file in os.listdir():
    if not (".csv" in file):
        continue
    abspath = os.path.abspath(file)
    temp_df = pd.read_csv(abspath, index_col = 0)
    
    player_salary = pd.concat([player_salary, temp_df], axis = 0, ignore_index=True) 

In [25]:
player_salary

선수    연도    팀   연봉(만원)    WAR
0       헥터  2017  KIA  170,000  5.803
1      양현종  2017  KIA  150,000  4.561
2      최형우  2017  KIA  150,000  6.674
3      윤석민  2017  KIA  125,000    NaN
4     버나디나  2017  KIA   75,000  5.336
...    ...   ...  ...      ...    ...
4623   유상빈  2023   한화    3,000 -0.092
4624   원혁재  2023   한화    3,000    NaN
4625   오세훈  2023   한화    3,000    NaN
4626   김현우  2023   한화    2,700    NaN
4627   김범준  2023   한화        0    NaN

[4628 rows x 5 columns]

In [26]:
# pitcher의 player_ID 컬럼 추가

df_3= pd.merge(player_salary, total_KBO_final, left_on='선수', right_on='pitcher', how='left')
df_3= df_3[['연도','팀', '선수','pitcher_ID', '연봉(만원)']].drop_duplicates()
df_3.reset_index().drop(columns = ['index'], axis=1)
df_3['pitcher_ID'].isnull().sum() # pitcher_ID 가 없는 선수는 1276명 / 없는 이유? 타자일 가능성!

# batter의 player_ID 컬럼 추가
df_4= pd.merge(player_salary, total_KBO_final, left_on='선수', right_on='batter', how='left')
df_4= df_4[['연도','팀', '선수','batter_ID', '연봉(만원)']].drop_duplicates()
df_4['batter_ID'].isnull().sum()

2892

In [27]:
df_3.rename(columns = {'pitcher_ID':'playerID'}, inplace=True)
df_3

연도    팀    선수  playerID   연봉(만원)
0       2017  KIA    헥터   66643.0  170,000
1614    2017  KIA   양현종   77637.0  150,000
4689    2017  KIA   최형우       NaN  150,000
4690    2017  KIA   윤석민   75620.0  125,000
4871    2017  KIA  버나디나       NaN   75,000
...      ...  ...   ...       ...      ...
951344  2023   한화   유상빈       NaN    3,000
951345  2023   한화   원혁재       NaN    3,000
951346  2023   한화   오세훈       NaN    3,000
951347  2023   한화   김현우   60457.0    2,700
951399  2023   한화   김범준       NaN        0

[4695 rows x 5 columns]

In [28]:
df_4.rename(columns = {'batter_ID':'playerID'}, inplace=True)
df_4

연도    팀    선수  playerID   연봉(만원)
0        2017  KIA    헥터   66643.0  170,000
1        2017  KIA   양현종       NaN  150,000
2        2017  KIA   최형우   72443.0  150,000
2389     2017  KIA   윤석민   74215.0  125,000
3748     2017  KIA  버나디나   67650.0   75,000
...       ...  ...   ...       ...      ...
1069752  2023   한화   유상빈       NaN    3,000
1069753  2023   한화   원혁재       NaN    3,000
1069754  2023   한화   오세훈       NaN    3,000
1069755  2023   한화   김현우       NaN    2,700
1069756  2023   한화   김범준       NaN        0

[4747 rows x 5 columns]

In [29]:
# concatenate
player_salary_info = pd.concat([df_3, df_4],axis=0, ignore_index=True) # 총 1171명의 선수정보 2017-2023
player_salary_info.isnull().sum()

연도             0
팀              0
선수             0
playerID    5872
연봉(만원)         0
dtype: int64

In [30]:
player_salary_info

연도    팀    선수  playerID   연봉(만원)
0     2017  KIA    헥터   66643.0  170,000
1     2017  KIA   양현종   77637.0  150,000
2     2017  KIA   최형우       NaN  150,000
3     2017  KIA   윤석민   75620.0  125,000
4     2017  KIA  버나디나       NaN   75,000
...    ...  ...   ...       ...      ...
9437  2023   한화   유상빈       NaN    3,000
9438  2023   한화   원혁재       NaN    3,000
9439  2023   한화   오세훈       NaN    3,000
9440  2023   한화   김현우       NaN    2,700
9441  2023   한화   김범준       NaN        0

[9442 rows x 5 columns]

In [31]:
player_salary_info = player_salary_info.dropna(subset = ['playerID'], axis = 0)
player_salary_info = player_salary_info.reset_index().drop(columns=['index'], axis = 1)
player_salary_info['playerID'] = player_salary_info['playerID'].astype('int')
player_salary_info

연도    팀   선수  playerID   연봉(만원)
0     2017  KIA   헥터     66643  170,000
1     2017  KIA  양현종     77637  150,000
2     2017  KIA  윤석민     75620  125,000
3     2017  KIA   팻딘     67645   70,000
4     2017  KIA  임창용     95657   50,000
...    ...  ...  ...       ...      ...
3565  2023   한화  박정현     50709    3,500
3566  2023   한화  이도윤     65703    3,400
3567  2023   한화  김인환     66715    3,200
3568  2023   한화  최원준     66606    3,000
3569  2023   한화  박준영     66928    3,000

[3570 rows x 5 columns]

In [32]:
final_salary_info = player_salary_info.copy()

In [33]:
os.getcwd()

'C:\\Users\\nhw85\\Downloads\\[KBO] 팀별 선수 연봉(17-23)'

In [34]:
final_salary_info.to_csv('final_salary_info' + '2017_2023' + '.csv', encoding = 'utf-8')

#### 몇몇 선수들의 데이터가 없는 것 같다..! 
ex. personal_player_info 에는 없는 선수가 player_salary 에는 있는 선수가 있다...

#### master.csv

In [39]:
### pitcher ID list 
pit_ID = []
pit_retro_ID = []
for pitcher_ID in total_KBO_final["pitcher_ID"]:
    if pitcher_ID == '':
        pass
    if pitcher_ID not in pit_ID:
        pit_ID.append(pitcher_ID)
        pit_retro_ID.append(str(pitcher_ID) + 'abc')


In [42]:
### batter ID list 
bat_ID = []
bat_retro_ID = []
for batter_ID in total_KBO_final["batter_ID"]:
    if batter_ID == '':
        pass
    if batter_ID not in bat_ID:
        bat_ID.append(batter_ID)
        bat_retro_ID.append(str(batter_ID) + 'abc')

In [47]:
playerID = pit_ID + bat_ID
retroID = pit_retro_ID + bat_retro_ID

In [48]:
master = pd.DataFrame()
master['playerID'] = playerID
master['retroID'] = retroID


In [49]:
master

playerID   retroID
0       61240  61240abc
1       67742  67742abc
2       74731  74731abc
3       99737  99737abc
4       79764  79764abc
..        ...       ...
951     67063  67063abc
952     50868  50868abc
953     50996  50996abc
954     69706  69706abc
955     68504  68504abc

[956 rows x 2 columns]

In [50]:
os.getcwd()

'C:\\Users\\nhw85\\Downloads\\[KBO] 팀별 선수 연봉(17-23)'

In [51]:
master.to_csv('master' + '2017_2023' + '.csv', encoding = 'utf-8')

In [ ]:
master = pd.read_csv("C://Users//nhw85//OneDrive//바탕 화면//master2017_2023.csv", index_col=0)
master 